pkg/build/reaper/reaper.go

```diff
@@ -0,0 +1,107 @@
+package reaper
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	kapi "k8s.io/kubernetes/pkg/api"
+	kerrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/kubectl"
+	kutilerrors "k8s.io/kubernetes/pkg/util/errors"
+
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	buildutil "github.com/openshift/origin/pkg/build/util"
+	"github.com/openshift/origin/pkg/client"
+	"github.com/openshift/origin/pkg/util"
+)
+
+// NewBuildConfigReaper returns a new reaper for buildConfigs
+func NewBuildConfigReaper(oc *client.Client) kubectl.Reaper {
+	return &BuildConfigReaper{oc: oc, pollInterval: kubectl.Interval, timeout: kubectl.Timeout}
+}
+
+// BuildConfigReaper implements the Reaper interface for buildConfigs
+type BuildConfigReaper struct {
+	oc                    client.Interface
+	pollInterval, timeout time.Duration
+}
+
+// Stop deletes the build configuration and all of the associated builds.
+func (reaper *BuildConfigReaper) Stop(namespace, name string, timeout time.Duration, gracePeriod *kapi.DeleteOptions) (string, error) {
+	noBcFound := false
+	noBuildFound := true
+
+	// Add deletion pending annotation to the build config
+	for {
+		bc, err := reaper.oc.BuildConfigs(namespace).Get(name)
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				noBcFound = true
+				break
+			}
+			return "", err
+		}
+		err = util.AddObjectAnnotations(bc, map[string]string{
+			buildapi.BuildConfigDeletionAnnotation: "",
```

Comment:
since we're going to allow the BC to stick around if something went wrong, but with the annotation in place, this logic needs to check if the annotation already exists and not add if it so, so we don't end up with it duplicated.